### PR TITLE
React Dropdown API Upgrades

### LINKED ISSUE #3496: Dropdown API Upgrades - Part 2 (React)
# Dropdown
Dropdowns use a config object to build the dropdown items. That method seems to be okay here for now because they aren't a highly customizable component (such as cards). For all work detailed below, please use the deprecation strategy process.

**Notes:**
SME: Amber
Part 2, Angular will be done in #3495 

## React
-- new for all API work, must support router links
- Change these prop names (following standards for deprecation strategy)
  `additionalIconClasses` -> `iconAdditionalClasses`
  `additionalTriggerClasses` -> `triggerAdditionalClasses`
  `additionalTriggerTextClasses` -> `triggerTextAdditionalClasses`
  `title` -> `heading`
- Remove empty `defaultProps`.
- Change component logic from `variant === 'base'` to `variant !== 'informational`' and remove the defaultProp for `variant`.
- We aren't spreading the "rest" of the supplied props onto the dropdown. Add that ability in. 
- Fix eslint errors
- We have a prop of `iconName` that has a default and is used for the base dropdrown but the iconName in the informational icon is hardcoded. Update to use `iconName`.
- Provide props for handler functions to be run when the dropdown methods run.

### PR DESCRIPTION
**Pull request by bobwise**
## What does this PR do?
1. Deprecates `additionalIconClasses` and replaces it with `iconAdditionalClasses`.
2. Deprecates `additionalTriggerClasses ` and replaces it with `triggerAdditionalClasses `.
3. Deprecates `additionalTriggerTextClasses ` and replaces it with `triggerTextAdditionalClasses `.
4. Deprecates `title ` and replaces it with `heading`.
5. Adds new events `openedEvent` and `closedEvent`.
5. Allows both variants to use `iconName` and `iconAdditionalClasses`.
5. Updates internal logic to check for `variant !== 'informational'` instead of `variant === 'base'`.
6. Removes empty `defaultProps`.
7. Spreads `...rest` correctly .

### Associated Issue
Fixes #3496 and https://github.com/sparkdesignsystem/spark-design-system/issues/3577

### Code
 - [x] Build Component in React
 - [x] Unit Testing in React with `npm run test` in `react/` (100% coverage, 100% passing)

### Browser Testing (current version and 1 prior)
  - [x] Google Chrome
  - [x] Google Chrome (Mobile)
  - [x] Mozilla Firefox
  - [x] Mozilla Firefox (Mobile)
  - [x] Microsoft Internet Explorer 11 (only this specific version of IE)
  - [x] Microsoft Edge
  - [x] Apple Safari
  - [x] Apple Safari (Mobile)
  
### Accessibility Testing 
Dropdown accessibility will be fixed in another Issue https://github.com/sparkdesignsystem/Backlog/issues/1325

### Deploy Preview Reviewed and Approved
 - [ ] Design Team (no changes to design)
 - [ ] Accessibility Team (https://github.com/sparkdesignsystem/Backlog/issues/1325)

### CODE REVIEW

File: react/src/components/dropdown/SprkDropdown.js [hunk 1]
```diff
@@ -127,14 +154,15 @@ class SprkDropdown extends Component {
...
             'sprk-c-Dropdown__trigger',
             { 'sprk-u-mrs': variant === 'informational' },
-            additionalTriggerClasses,
+            triggerAdditionalClasses,
           )}
           aria-expanded={isOpen}
           aria-haspopup="listbox"
           aria-label={screenReaderText || triggerText}
           data-analytics={analyticsString || 'undefined'}
           data-id={idString || 'undefined'}
```
Comment (guern1kn) [inline react/src/components/dropdown/SprkDropdown.js:163]: I know this wasn't part of the AC, but we should update this. If there isn't an `analyticsString` or `data-id` they shouldn't be added.

File: react/src/components/dropdown/SprkDropdown.test.js [hunk 2]
```diff
@@ -463,4 +463,63 @@ describe('SprkDropdown:', () => {
...
+    wrapper.find('.sprk-b-Link').simulate('click');
+    expect(myMock.mock.calls.length).toBe(0);
+
+    wrapper.instance().closeOnEsc({ key: 'Escape' });
+    expect(myMock.mock.calls.length).toBe(1);
+
+    wrapper.instance().closeOnEsc({ key: 'Escape' });
+    expect(myMock.mock.calls.length).toBe(1);
+  });
 });
```
Comment (guern1kn) [inline react/src/components/dropdown/SprkDropdown.test.js:525]: Let's add tests for the new prop names.

File: react/src/components/dropdown/SprkDropdown.js [hunk 3]
```diff
@@ -147,37 +175,37 @@ class SprkDropdown extends Component {
                   sprk-c-Icon--stroke-current-color
                   sprk-u-mls
                 "
```
Comment (guern1kn) [inline react/src/components/dropdown/SprkDropdown.js:177]: We should allow the `iconAddiontionalClasses` to be added to both the default and informational variants.

File: react/src/components/dropdown/SprkDropdown.js [hunk 4]
```diff
@@ -127,14 +154,15 @@ class SprkDropdown extends Component {
...
           aria-expanded={isOpen}
           aria-haspopup="listbox"
           aria-label={screenReaderText || triggerText}
           data-analytics={analyticsString || 'undefined'}
           data-id={idString || 'undefined'}
           onClick={this.toggleDropdownOpen}
+          {...rest}
         >
           {variant === 'informational' && (
             <>
```
Comment (guern1kn) [inline react/src/components/dropdown/SprkDropdown.js:168]: Update this to be `triggerTextAdditionalClasses`

File: react/src/components/dropdown/SprkDropdown.js [hunk 5]
```diff
@@ -233,7 +261,7 @@ class SprkDropdown extends Component {
                               choiceFunction(value);
                             }
                           }}
-                          {...rest}
+                          {...choicesRest}
                         >
                           <p className="sprk-b-TypeBodyOne">{content.title}</p>
                           <p className="sprk-b-TypeBodyTwo">
```
Comment (guern1kn) [inline react/src/components/dropdown/SprkDropdown.js:267]: If an `infoLine` isn't present in the `item.content` then an empty `<p>` still gets rendered. Let's add some checks to only render what's necessary.

**Review by guern1kn — CHANGES_REQUESTED** [review]
Nice work on this, Bob. Left a few comments. I feel like there's still quite a few elements of this that we could improve on customization, like adding extra classes. For example, `footerAdditionalClasses`. What are your thoughts on adding some more props like that?

**Comment by bobwise** [comment]
> Nice work on this, Bob. Left a few comments. I feel like there's still quite a few elements of this that we could improve on customization, like adding extra classes. For example, `footerAdditionalClasses`. What are your thoughts on adding some more props like that?

Thanks @guern1kn! I agree there are a lot of parts of this component that can't currently be customized. There are comments in the original Issue that indicate that we are comfortable with the current level of customization.

> Dropdowns use a config object to build the dropdown items. That method seems to be okay here for now because they aren't a highly customizable component (such as cards).

This would be a good thing to bring up to the whole group I think. I'd rather refactor it into a compositional component than try to add 2-5 new `additionalClasses` props at this point.

**Review by guern1kn — APPROVED** [review]
lgtm

**Review by afebbraro — CHANGES_REQUESTED** [review]
Hey @bobwise , with the router issue recently, we need to make sure all our components support router links. I tried with this and it doesnt work, i get this error Warning: Failed prop type: Invalid prop `choices.items[0].element` supplied to `SprkDropdown`.

I updated the AC for this component. Should hopefully be an easy addition to this PR.

**Review by afebbraro — APPROVED** [review]
Looks great, thank you for incorporating the browser router.

**Review by guern1kn — APPROVED** [review]

**Pull request merged**